File: dev-repo/opportunityplus-260506_dev-deploy/UNOPS.PAO.Models/Interactions/InteractionModel.cs
using UNOPS.PAO.Domain.Enums;
using System.Text.Json.Serialization;
using System.Text;
using UNOPS.PAO.Domain.Entities;
using UNOPS.PAO.Models.OrganizationUnits;
using UNOPS.PAO.Models.Partners;
using UNOPS.PAO.Models.Contacts;
using UNOPS.PAO.Models.Users;
using UNOPS.PAO.Models.Documents;
using UNOPS.PAO.Models.Shared;

namespace UNOPS.PAO.Models.Interactions;

public class InteractionModel
{
    public int Id { get; set; }
    public InteractionType Type { get; set; }
    public DateTime Date { get; set; }

    public string? Description { get; set; }
    
    public string? ContactName { get; set; }
    public string Status { get; set; }
    public virtual List<string>? EmailAddresses { get; set; } = new List<string>();
    public List<int>? ContactIds { get; set; } = new List<int>();
    [JsonIgnore]
    public virtual ICollection<InteractionContactModel>? InteractionContacts { get; set; }
    public List<int>? PartnerIds { get; set; } = new List<int>();
    [JsonIgnore]
    public virtual ICollection<InteractionPartnerModel>? InteractionPartners { get; set; }
    [JsonIgnore]
    public virtual ICollection<InteractionUserModel>? InteractionUsers { get; set; }
    public string? Location { get; set; }
    public string Subject { get; set; }
    
    /// <summary>Interaction office scope (from <c>OfficeRelationship</c>), serialized as <c>officeRelationships</c>.</summary>
    [JsonPropertyName("officeRelationships")]
    public List<OrganizationUnitRelationshipModel>? OfficeRelationships { get; set; }
    
    public List<DocumentModel>? Documents { get; set; }

    /// <summary>
    /// Full contact entities associated with this interaction
    /// </summary>
    public List<ContactModel>? Contacts { get; set; }

    /// <summary>
    /// Full partner entities associated with this interaction
    /// </summary>
    public List<PartnerModel>? Partners { get; set; }

    /// <summary>
    /// Full user entities associated with this interaction
    /// </summary>
    public List<UserValueModel>? Users { get; set; }

    /// <summary>
    /// Permissions for this specific interaction
    /// </summary>
    public EntityPermissionsModel? Permissions { get; set; }
    public string? GmailThreadId { get; set; }
    public string? GmailMessageId { get; set; }
    
    // Computed properties from NotMapped fields in Interaction entity
    public string? InteractionContactsList { get; set; }
    public string? InteractionPartnersList { get; set; }
    public string? InteractionUsersList { get; set; }
    public string? InteractionOrgUnits { get; set; }
    
    // Audit fields from ModifiableDeletableEntity (read-only from frontend perspective)
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public DateTime? CreatedDate { get; set; }
    
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
    public DateTime? LastModifiedDate { get; set; }
    
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? CreatedBy { get; set; }
    
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? LastModifiedBy { get; set; }
    
    // User name fields resolved from UserProfile
    public string? CreatedByName { get; set; }
    public string? LastModifiedByName { get; set; }
}

public class InteractionContactModel
{
    public int InteractionId { get; set; }
    public int ContactId { get; set; }
}

public class InteractionPartnerModel
{
    public int InteractionId { get; set; }
    public int PartnerId { get; set; }
}

public class InteractionUserModel
{
    public int InteractionId { get; set; }
    public int UserId { get; set; }
}

File: dev-repo/opportunityplus-260506_dev-deploy/UNOPS.PAO.ClientApp/src/app/features/partnerships/interactions/models/interaction.model.ts
import { InteractionType } from './interaction-type.enum';
import { EntityPermissionSet } from '@shared/models/shared-types';
import { OrganizationUnitRelationshipModel } from '@partnerships/partners/models/organization-unit-relationship.model';
import { Contact } from '@partnerships/contacts/models/contact.model';
import { Partner } from '@partnerships/partners/models/partner.model';
import { UserValueModel } from '@shared/models/user.model';

export interface DocumentModel {
  id: number;
  name: string;
  size?: number;
  type?: string;
  url?: string;
}

export interface Interaction {
  id: number;
  type: InteractionType;
  date: string;
  description?: string;
  contactId: number;
  contactName?: string;
  status: string;
  contactIds: number[];
  partnerIds: number[];
  emailAddresses: string[];
  location: string;
  subject: string;
  /** Office links (API); legacy key was organizationUnitRelationships */
  officeRelationships?: OrganizationUnitRelationshipModel[] | null;
  organizationUnitRelationships?: OrganizationUnitRelationshipModel[] | null;
  createdBy: number;
  permissions?: EntityPermissionSet;
  
  // Documents/Attachments
  documents?: DocumentModel[];
  
  // Full related entities (from backend)
  contacts?: Contact[];
  partners?: Partner[];
  users?: UserValueModel[];
  
  // Gmail integration
  gmailThreadId?: string;
  gmailMessageId?: string;
  
  // Audit fields
  createdDate?: string;
  lastModifiedDate?: string;
  lastModifiedBy?: number;
  
  // Resolved user names for audit fields
  createdByName?: string;
  lastModifiedByName?: string;
  
  /** Comma-joined display of linked org units (API computed). */
  interactionOrgUnits?: string | null;

  // Import-specific properties
  isImportEdit?: boolean;
  _updated?: boolean;
  _importRowId?: string;
}

/** Resolves office/org scope from API payload (prefers `officeRelationships`). */
export function getInteractionOfficeRelationships(
  interaction: Interaction | null | undefined
): OrganizationUnitRelationshipModel[] | null | undefined {
  return interaction?.officeRelationships ?? interaction?.organizationUnitRelationships;
}
